lib/compiler_rt/hexagon.zig

```diff
@@ -0,0 +1,611 @@
+        \\   r2 = lsl(r1,r3)
+        \\  }
+        \\  .falign
+        \\ 1:
+        \\  {
+        \\   p0 = cmp.gtu(r2,r0)
+        \\   if (!p0.new) r0 = sub(r0,r2)
+        \\   r2 = lsr(r2,#1)
+        \\   if (p1) r1 = #0
+        \\  }:endloop0
+        \\  {
+        \\   p0 = cmp.gtu(r2,r0)
+        \\   if (!p0.new) r0 = sub(r0,r1)
+        \\   if (p2) jumpr r31
+        \\  }
+        \\  {
+        \\   r0 = neg(r0)
+        \\   jumpr r31
+        \\  }
+    );
+}
+
+fn __hexagon_memcpy_likely_aligned_min32bytes_mult8bytes() callconv(.naked) noreturn {
+    asm volatile (
+        \\ {
+        \\   p0 = bitsclr(r1,#7)
+        \\   p0 = bitsclr(r0,#7)
+        \\   if (p0.new) r5:4 = memd(r1)
+        \\   r3 = #-3
+        \\  }
+        \\  {
+        \\   if (!p0) jump .Lmemcpy_call
+        \\   if (p0) memd(r0++#8) = r5:4
+        \\   if (p0) r5:4 = memd(r1+#8)
+        \\   r3 += lsr(r2,#3)
+        \\  }
+        \\  {
+        \\   memd(r0++#8) = r5:4
+        \\   r5:4 = memd(r1+#16)
+        \\   r1 = add(r1,#24)
+        \\   loop0(1f,r3)
+        \\  }
+        \\  .falign
+        \\ 1:
+        \\  {
+        \\   memd(r0++#8) = r5:4
+        \\   r5:4 = memd(r1++#8)
+        \\  }:endloop0
+        \\  {
+        \\   memd(r0) = r5:4
+        \\   r0 -= add(r2,#-8)
+        \\   jumpr r31
+        \\  }
+        \\ .Lmemcpy_call:
+        \\      jump memcpy@PLT
+    );
+}
+
+fn __hexagon_udivsi3() callconv(.naked) noreturn {
+    asm volatile (
+        \\ {
+        \\   r2 = cl0(r0)
+        \\   r3 = cl0(r1)
+        \\   r5:4 = combine(#1,#0)
+        \\   p0 = cmp.gtu(r1,r0)
+        \\  }
+        \\  {
+        \\   r6 = sub(r3,r2)
+        \\   r4 = r1
+        \\   r1:0 = combine(r0,r4)
+        \\   if (p0) jumpr r31
+        \\  }
+        \\  {
+        \\   r3:2 = vlslw(r5:4,r6)
+        \\   loop0(1f,r6)
+        \\  }
+        \\  .falign
+        \\ 1:
+        \\  {
+        \\   p0 = cmp.gtu(r2,r1)
+        \\   if (!p0.new) r1 = sub(r1,r2)
+        \\   if (!p0.new) r0 = add(r0,r3)
+        \\   r3:2 = vlsrw(r3:2,#1)
+        \\  }:endloop0
+        \\  {
+        \\   p0 = cmp.gtu(r2,r1)
+        \\   if (!p0.new) r0 = add(r0,r3)
+        \\   jumpr r31
+        \\  }
+    );
+}
+
+comptime {
+    if (builtin.cpu.arch == .hexagon) {
+        @export(__hexagon_divsi3, .{ .name = "__hexagon_divsi3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_udivmoddi4, .{ .name = "__hexagon_udivmoddi4", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_umodsi3, .{ .name = "__hexagon_umodsi3", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_sqrtf, .{ .name = "__hexagon_sqrtf", .linkage = common.linkage, .visibility = common.visibility });
+        @export(__hexagon_udivmodsi4, .{ .name = "__hexagon_udivmodsi4", .linkage = common.linkage, .visibility = common.visibility });
```

Comment:
Is this one actually used by LLVM?

Reply:
No, I don't believe so.  I included it in order to preserve the published builtins from `llvm-project/compiler-rt` but I'm happy to omit it.

Reply:
I think we can just omit it then. We only really care about those that the compiler actually uses in practice.